#### pd.melt arguments
- id_vars = columns you want to keep (not melt)
- var_name = name of new column you created by melting columns
- value_name = column name for values

#### pd.pivot_table arguments
- Index = columns you want to keep (not pivot)
- columns = column you want to pivot
- values = values we want to populate in the new columns
- aggfunc = how you want to aggregate the duplicate rows

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pydataset import data
import statistics
import seaborn as sns
import env

import acquire

# 1. Attendance Data

### Load the attendance.csv file and calculate an attendnace percentage for each student. 

### One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [2]:
# You should end up with something like this:

# name
# Billy    0.5250
# Jane     0.6875
# John     0.9125
# Sally    0.7625
# Name: grade, dtype: float64


# p = 1
# t = .9
# h = .5

In [3]:
attendance = pd.read_csv('attendance.csv', index_col=0)
attendance.head()

2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
Sally          P          T          T          H          P          A   
Jane           A          P          T          T          T          T   
Billy          A          T          A          A          H          T   
John           P          T          H          P          P          T   

      2018-01-07 2018-01-08  
Sally          T          T  
Jane           A          T  
Billy          P          T  
John           P          P

In [4]:
attendance.shape

(4, 8)

In [5]:
(1 + .9 + .5 + 1 + 1 + .9 + 1 + 1) / 8
#this would represent john's attendance

0.9125000000000001

### the goal here will be to get this into `name` `date` `attendance` format

In [6]:
attendance.index.rename('student_name', inplace=True)
attendance

#create a title for the name column

2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
student_name                                                          
Sally                 P          T          T          H          P   
Jane                  A          P          T          T          T   
Billy                 A          T          A          A          H   
John                  P          T          H          P          P   

             2018-01-06 2018-01-07 2018-01-08  
student_name                                   
Sally                 A          T          T  
Jane                  T          A          T  
Billy                 T          P          T  
John                  T          P          P

In [7]:
attendance = attendance.reset_index()
attendance

#reset the index

student_name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0        Sally          P          T          T          H          P   
1         Jane          A          P          T          T          T   
2        Billy          A          T          A          A          H   
3         John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [8]:
attendance_melt = attendance.melt(id_vars='student_name', var_name='date', value_name='attendance')


# melt the many columns into their vars and vals.  date and attendance here.  kept 'student_name'

In [9]:
attendance_melt.sample(5)

student_name        date attendance
6         Billy  2018-01-02          T
22        Billy  2018-01-06          T
23         John  2018-01-06          T
7          John  2018-01-02          T
2         Billy  2018-01-01          A

In [10]:
attendance_melt.dtypes


student_name    object
date            object
attendance      object
dtype: object

In [11]:
# p = 1
# t = .9
# h = .5

# P = attendance_melt[(attendance_melt.attendance == 'P')]
# T = attendance_melt[(attendance_melt.attendance == 'T')]
# H = attendance_melt[(attendance_melt.attendance == 'H')]

attendance_melt.head()

# the above was unecessary in the end

student_name        date attendance
0        Sally  2018-01-01          P
1         Jane  2018-01-01          A
2        Billy  2018-01-01          A
3         John  2018-01-01          P
4        Sally  2018-01-02          T

In [12]:
attendance_clean = attendance_melt.replace({'P':1, 'T': .9, 'H': .5, 'A':0})
attendance_clean.head()

# swapping those alpha values for their corresponding numeric ones for ease of calculation below

student_name        date  attendance
0        Sally  2018-01-01         1.0
1         Jane  2018-01-01         0.0
2        Billy  2018-01-01         0.0
3         John  2018-01-01         1.0
4        Sally  2018-01-02         0.9

In [13]:
attendance_clean.groupby('student_name').mean()

# and.... the avg of every student

attendance
student_name            
Billy             0.5250
Jane              0.6875
John              0.9125
Sally             0.7625

In [14]:
#good work

# 2. Coffee Levels

### Read the coffee_levels.csv file.
### Transform the data so that each carafe is in it's own column.


In [15]:
coffee = pd.read_csv('coffee_levels.csv', index_col=0)
coffee.head()

coffee_carafe  coffee_amount
hour                             
8                x       0.816164
9                x       0.451018
10               x       0.843279
11               x       0.335533
12               x       0.898291

In [16]:
coffee.shape

(30, 2)

In [17]:
coffee_tidy = coffee.pivot_table(index = 'hour',
                   columns = 'coffee_carafe',
                   values = 'coffee_amount').reset_index()
coffee_tidy

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

### Is this the best shape for the data?

In [18]:
# i think the data was better in its original form, for analytical purposes at least

#### pd.pivot_table arguments
- Index = columns you want to keep (not pivot)
- columns = column you want to pivot
- values = values we want to populate in the new columns
- aggfunc = how you want to aggregate the duplicate rows

# 3. Cake Recipes

### Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.
### Tidy the data as necessary.

In [19]:
cake = pd.read_csv('cake_recipes.csv', index_col=0).reset_index()
cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [20]:
cake_melt = cake.melt(id_vars='recipe:position', var_name='temps',value_name='rating')
cake_melt.head()

# here i melted the oven temp setting into a variable and made the ratings the values

recipe:position temps     rating
0        a:bottom   225  61.738655
1           a:top   225  51.709751
2        b:bottom   225  57.095320
3           b:top   225  82.455004
4        c:bottom   225  96.470207

#### pd.melt arguments
- id_vars = columns you want to keep (not melt)
- var_name = name of new column you created by melting columns
- value_name = column name for values

In [21]:
# now, toca hacer un nuevo DF, separando los dos valores de la columna 'recipe_position'
# i'm following the pattern from the lesson example below

recipe_position_df = cake_melt['recipe:position'].str.split(':', expand = True)
recipe_position_df.columns = ['recipe', 'position']
recipe_position_df.head()

# year_and_measure_df = sales_melt.year_and_measure.str.split(' ', expand=True)
# year_and_measure_df.columns = ['year', 'measure']
# year_and_measure_df.head()

recipe position
0      a   bottom
1      a      top
2      b   bottom
3      b      top
4      c   bottom

In [22]:
# now, making new df to concatenat the previous two

cake_tidy = pd.concat([cake_melt,recipe_position_df], axis = 1).drop(columns='recipe:position')
cake_tidy.head()


#copying from the lesson here:
# sales2 = pd.concat([sales_melt, year_and_measure_df], axis=1).drop(columns='year_and_measure')
# sales2.head()

temps     rating recipe position
0   225  61.738655      a   bottom
1   225  51.709751      a      top
2   225  57.095320      b   bottom
3   225  82.455004      b      top
4   225  96.470207      c   bottom

In [23]:
# actually, this is sufficient for our purposes now

### Which recipe, on average, is the best? recipe b

In [43]:


score = cake_tidy.groupby('recipe').rating.mean().max()
score

76.73607424644435

In [45]:
# .idxmax ftw!!

best_recipe = cake_tidy.groupby('recipe').rating.mean().idxmax()

print(f' The best recipe is {best_recipe} with a score of {score}')

 The best recipe is b with a score of 76.73607424644435


### Which oven temperature, on average, produces the best results? 275

In [32]:
cake_tidy.groupby('temps').rating.mean().idxmax()

'275'

### Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [31]:
cake_tidy.groupby(by = ['recipe','position','temps']).rating.mean().idxmax()

('b', 'bottom', '300')

In [33]:
cake_tidy.groupby(by = ['recipe','position','temps']).rating.mean()

recipe  position  temps
a       bottom    225      61.738655
                  250      53.912627
                  275      74.414730
                  300      98.786784
        top       225      51.709751
                  250      52.009735
                  275      68.576858
                  300      50.228470
b       bottom    225      57.095320
                  250      61.904369
                  275      61.196980
                  300      99.248541
        top       225      82.455004
                  250      95.224151
                  275      98.594881
                  300      58.169349
c       bottom    225      96.470207
                  250      52.001358
                  275      92.893227
                  300      65.473084
        top       225      71.306308
                  250      82.795477
                  275      92.098049
                  300      53.960273
d       bottom    225      52.799753
                  250      58.670419
              